File: 202-happy-number/happy-number.go
var se = map[int]bool {1:true}

func isHappy(n int) bool {
if v,exists := se[n]; exists{
    return v;
}else{
    sos := 0
    no := n
    for n > 0 {
        d := n % 10
        sos += d * d
        n/= 10
    }
     se[no] = false
     se[no] = isHappy(sos)
     return se[no]
}

}

File: 202-happy-number/happy-number.cpp
class Solution {
public:
    int sos(int x){
        int sum = 0;
        while(x){
            int d =  x % 10;
            x/=10;
            sum += d * d;
        }
        return sum;
    }
    map<int,bool> mp = {{1,true}};
    bool isHappy(int n) {
        if(n==1) return true;
        if(mp.count(n))return mp[n];
        mp[n] = false;
       return mp[n] = isHappy(sos(n));
    }

};